### PR TITLE
[REVIEW] Add builder API for cuIO `orc_writer_options` and `orc_reader_options`

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
#5962

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Review by ajschmidt8 — APPROVED** [review]

**Review by harrism — CHANGES_REQUESTED** [review]
Looks good, minor comments.

File: cpp/include/cudf/io/orc.hpp [hunk 1]
```diff
@@ -0,0 +1,834 @@
...
+    CUDF_EXPECTS(nrows == -1 or _stripes.empty(), "Can't set both num_rows along with stripes");
+    _num_rows = (nrows != 0) ? nrows : -1;
+  }
+
+  /**
+   * @brief Enable/Disable use of row index to speed-up reading.
+   *
+   * @param use Boolean value to enable/disable row index use.
+   */
+  void set_use_index(bool use) { _use_index = use; }
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/orc.hpp:203]: Missed it for set, nice catch.

File: cpp/benchmarks/io/orc/orc_reader_benchmark.cpp [hunk 2]
```diff
@@ -51,10 +51,14 @@ void ORC_read(benchmark::State& state)
   auto const tbl  = create_random_table<T>(num_cols, col_bytes, true);
   auto const view = tbl->view();
 
-  cudf_io::write_orc_args args{cudf_io::sink_info(&out_buffer), view, nullptr, compression};
+  cudf_io::orc_writer_options args =
+    cudf_io::orc_writer_options::builder(cudf_io::sink_info(&out_buffer), view)
+      .metadata(nullptr)
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/benchmarks/io/orc/orc_reader_benchmark.cpp:56]: I just updated it in case default value might change, but will do that.

File: cpp/include/cudf/io/detail/orc.hpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file readers.hpp
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/detail/orc.hpp:18]: We need to, this helps in mentioning header file in docs.

File: cpp/include/cudf/io/detail/orc.hpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019i-2020, NVIDIA CORPORATION.
```
Comment (davidwendt) [inline cpp/include/cudf/io/detail/orc.hpp:2]: If this is a new file then it does not need 2019. Also there is an extra 'i' in there.

File: cpp/include/cudf/io/orc.hpp [hunk 5]
```diff
@@ -0,0 +1,832 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file functions.hpp
```
Comment (davidwendt) [inline cpp/include/cudf/io/orc.hpp:18]: File name does not match

File: cpp/include/cudf/io/orc.hpp [hunk 6]
```diff
@@ -0,0 +1,832 @@
...
+ *    ...
+ *  cudf_write_orc_chunked_end(state);
+ * @endcode
+ *
+ * @param[in] options Settings for controlling writing behavior.
+ * @param[in] mr Device memory resource to use for device memory allocation.
+ *
+ * @returns pointer to an anonymous state structure storing information about the chunked write.
+ * this pointer must be passed to all subsequent write_orc_chunked() and write_orc_chunked_end()
+ *          calls.
```
Comment (davidwendt) [inline cpp/include/cudf/io/orc.hpp:801]:
```suggestion
 * calls.
```

**Review by davidwendt — CHANGES_REQUESTED** [review]
Looks good. Just some minor items.

File: cpp/include/cudf/io/orc.hpp (hunk 5 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/orc.hpp:18]: nice catch

File: cpp/include/cudf/io/detail/orc.hpp (hunk 4 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/detail/orc.hpp:2]: But these functions were copied from another file as is, so wasn't sure about that.

Comment (davidwendt) [reply cpp/include/cudf/io/detail/orc.hpp]: If the file was just renamed keeping the old date would be ok but this is really a new file. The old file will still have its own history.

Comment (rgsl888prabhu) [reply cpp/include/cudf/io/detail/orc.hpp]: I have made the changes.

**Review by davidwendt — APPROVED** [review]

**Comment by rgsl888prabhu** [comment]
rerun tests

File: python/cudf/cudf/_lib/orc.pyx [hunk 7]
```diff
@@ -101,43 +103,60 @@ cpdef write_orc(Table table,
...
-    cdef write_orc_args c_write_orc_args = write_orc_args(
-        sink_info_c,
-        table.data_view(), &metadata_, compression_,
-        <bool> (True if enable_statistics else False)
+    cdef orc_writer_options c_orc_writer_options = move(
+        orc_writer_options.builder(sink_info_c, table.data_view()).
+        metadata(&metadata_).
+        compression(compression_).
+        enable_statistics(<bool> (True if enable_statistics else False)).
+        build()
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/orc.pyx:111]:
```suggestion
        orc_writer_options.builder(sink_info_c, table.data_view())
        .metadata(&metadata_)
        .compression(compression_)
        .enable_statistics(<bool> (True if enable_statistics else False))
        .build()
```

File: python/cudf/cudf/_lib/orc.pyx [hunk 8]
```diff
@@ -101,43 +103,60 @@ cpdef write_orc(Table table,
...
 
 
 cdef size_type get_size_t_arg(arg, name) except*:
-    arg = -1 if arg is None else arg
-    if not isinstance(arg, int) or arg < -1:
-        raise TypeError("{} must be an int >= -1".format(name))
+    if name == "skip_rows":
+        arg = 0 if arg is None else arg
+        if not isinstance(arg, int) or arg < 0:
+            raise TypeError("{} must be an int >= 0".format(name))
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/orc.pyx:122]:
```suggestion
            raise TypeError(f"{name} must be an int >= 0")
```

File: python/cudf/cudf/_lib/orc.pyx [hunk 9]
```diff
@@ -101,43 +103,60 @@ cpdef write_orc(Table table,
...
+    else:
+        arg = -1 if arg is None else arg
+        if not isinstance(arg, int) or arg < -1:
+            raise TypeError("{} must be an int >= -1".format(name))
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/orc.pyx:126]:
```suggestion
            raise TypeError(f"{name} must be an int >= -1")
```

File: python/cudf/cudf/_lib/orc.pyx [hunk 10]
```diff
@@ -101,43 +103,60 @@ cpdef write_orc(Table table,
...
+        orc_reader_options.builder(src).
+        columns(c_column_names).
+        stripes(strps).
+        skip_rows(skip_rows).
+        num_rows(num_rows).
+        timestamp_type(data_type(timestamp_type)).
+        use_index(use_index).
+        decimals_as_float64(decimals_as_float).
+        forced_decimals_scale(force_decimal_scale).
+        build()
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/orc.pyx:159]:
```suggestion
        orc_reader_options.builder(src)
        .columns(c_column_names)
        .stripes(strps)
        .skip_rows(skip_rows)
        .num_rows(num_rows)
        .timestamp_type(data_type(timestamp_type))
        .use_index(use_index)
        .decimals_as_float64(decimals_as_float)
        .forced_decimals_scale(force_decimal_scale)
        .build()
```

**Review by kkraus14 — CHANGES_REQUESTED** [review]

**Review by kkraus14 — APPROVED** [review]

File: cpp/include/cudf/io/orc.hpp [hunk 11]
```diff
@@ -0,0 +1,832 @@
...
+
+  /**
+   * @brief Sets list of individual stripes to read.
+   *
+   * @param stripes Vector of individual stripes.
+   */
+  void set_stripes(std::vector<size_type> stripes)
+  {
+    CUDF_EXPECTS(stripes.empty() or (_skip_rows == 0 and _num_rows == -1),
+                 "Can't set both stripes along with skip_rows/num_rows");
```
Comment (vuule) [inline cpp/include/cudf/io/orc.hpp:170]: This would improve the error message
```suggestion
    CUDF_EXPECTS(stripes.empty() or (_skip_rows == 0), "Can't set stripes along with skip_rows");
    CUDF_EXPECTS(stripes.empty() or (_num_rows == -1),  "Can't set stripes along with num_rows");
```

File: cpp/include/cudf/io/orc.hpp [hunk 12]
```diff
@@ -0,0 +1,832 @@
...
+
+  /**
+   * @brief Sets number of row to read.
+   *
+   * @param nrows Number of rows.
+   */
+  void set_num_rows(size_type nrows)
+  {
+    CUDF_EXPECTS(nrows == -1 or _stripes.empty(), "Can't set both num_rows along with stripes");
+    _num_rows = (nrows != 0) ? nrows : -1;
```
Comment (vuule) [inline cpp/include/cudf/io/orc.hpp:193]: Why is this line required? Parquet reader options don't have this logic.

Comment (rgsl888prabhu) [reply cpp/include/cudf/io/orc.hpp:193]: To take care of this condition https://github.com/rapidsai/cudf/blob/c7e8a5b431ff03bb456c8aa54456e40c19c73799/cpp/src/io/orc/reader_impl.cu#L852

Comment (vuule) [reply cpp/include/cudf/io/orc.hpp:193]: Looks like the behavior is to read all rows when num_rows is set to zero. I don't think we want this. 
AFAIK this is also true for the Parquet reader, so we probably want to open a separate issue for this. Thoughts?

Comment (rgsl888prabhu) [reply cpp/include/cudf/io/orc.hpp:193]: Agree, will remove this and see if any thing breaks.

Comment (vuule) [reply cpp/include/cudf/io/orc.hpp:193]: Just checked, and Parquet reads zero rows with zero num_rows (which is correct). So it would be great to align ORC reader in this PR.

In nothing breaks, add a test that would :)

File: cpp/include/cudf/io/detail/orc.hpp [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+ public:
+  /**
+   * @brief Constructor from an array of file paths
+   *
+   * @param filepaths Paths to the files containing the input dataset
+   * @param options Settings for controlling reading behavior
+   * @param mr Device memory resource to use for device memory allocation
+   */
+  explicit reader(std::vector<std::string> const& filepaths,
```
Comment (cwharris) [inline cpp/include/cudf/io/detail/orc.hpp:46]: Eventually we'll want to remove file path based constructors completely and use `datasource` exclusively, but I doubt it's convenient for this PR.

File: cpp/include/cudf/io/orc.hpp [hunk 14]
```diff
@@ -0,0 +1,832 @@
...
+ */
+
+/**
+ * @file orc.hpp
+ * @brief cuDF-IO freeform API
+ */
+
+#pragma once
+
+#include "types.hpp"
```
Comment (cwharris) [inline cpp/include/cudf/io/orc.hpp:24]: Looks like `types.hpp` is in the include dir.

```suggestion
#include <cudf/io/types.hpp>
```

File: cpp/include/cudf/io/orc.hpp [hunk 15]
```diff
@@ -0,0 +1,832 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file orc.hpp
+ * @brief cuDF-IO freeform API
+ */
+
```
Comment (cwharris) [inline cpp/include/cudf/io/orc.hpp:21]:
```suggestion
```

File: cpp/include/cudf/io/orc.hpp [hunk 16]
```diff
@@ -0,0 +1,832 @@
...
+    CUDF_EXPECTS(nrows == -1 or _stripes.empty(), "Can't set both num_rows along with stripes");
+    _num_rows = nrows;
+  }
+
+  /**
+   * @brief Enable/Disable use of row index to speed-up reading.
+   *
+   * @param use Boolean value to enable/disable row index use.
+   */
+  void enable_use_index(bool use) { _use_index = use; }
```
Comment (cwharris) [inline cpp/include/cudf/io/orc.hpp:201]: If we want consistency with the `set` convention, we could prepend `should` to the underlying
```suggestion
  void set_should_use_index(bool use) { _should_use_index = use; }
```

File: cpp/include/cudf/io/orc.hpp [hunk 17]
```diff
@@ -0,0 +1,832 @@
...
+  size_type get_forced_decimals_scale() const { return _forced_decimals_scale; }
+
+  // Setters
+
+  /**
+   * @brief Sets names of the column to read.
+   *
+   * @param col_names Vector of column names.
+   */
+  void set_columns(std::vector<std::string> col_names) { _columns = std::move(col_names); }
```
Comment (cwharris) [inline cpp/include/cudf/io/orc.hpp:160]: IMO we should make these setters (and any mutable operations) private since the builder is a friend. This way the options cannot be mutated elsewhere. This, combined with the "build" function which returns a copy of the options, would make options effectively immutable outside of their respective builder.

File: cpp/include/cudf/io/orc.hpp [hunk 18]
```diff
@@ -0,0 +1,832 @@
...
+  table_metadata const* get_metadata() const { return _metadata; }
+
+  // Setters
+
+  /**
+   * @brief Sets compression type.
+   *
+   * @param comp Compression type.
+   */
+  void set_compression(compression_type comp) { _compression = comp; }
```
Comment (cwharris) [inline cpp/include/cudf/io/orc.hpp:482]: imho options setters / mutable functions should be private.

File: cpp/include/cudf/io/orc.hpp [hunk 19]
```diff
@@ -0,0 +1,832 @@
...
+  table_metadata_with_nullability const* get_metadata() const { return _metadata; }
+
+  // Setters
+
+  /**
+   * @brief Sets compression type.
+   *
+   * @param comp The compression type to use.
+   */
+  void set_compression(compression_type comp) { _compression = comp; }
```
Comment (cwharris) [inline cpp/include/cudf/io/orc.hpp:684]: prefer private setters and mutable functions for options.

File: cpp/include/cudf/io/orc.hpp [hunk 20]
```diff
@@ -0,0 +1,832 @@
...
+ * @ingroup io_writers
+ *
+ * All tables passed into multiple calls of this function must contain the same # of columns and
+ * have columns of the same type.
+ *
+ * @param[in] table The table data to be written.
+ * @param[in] state Opaque state information about the writer process. Must be the same pointer
+ * returned from write_orc_chunked_begin().
+ */
+void write_orc_chunked(table_view const& table, std::shared_ptr<orc_chunked_state> state);
```
Comment (cwharris) [inline cpp/include/cudf/io/orc.hpp:819]: should we make `state` be the first parameter, so it's consistent with `write_orc_chunked_end`?

**Review by cwharris — CHANGES_REQUESTED** [review]
Looks great. Can we make the options setters/mutable functions private so only their respective builders can access them?

File: cpp/include/cudf/io/orc.hpp (hunk 15 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/orc.hpp:21]: Why should this be removed ?

File: cpp/include/cudf/io/orc.hpp (hunk 16 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/orc.hpp:201]: There was a discussion where we decided to go with enable and is_eanbled for boolean options. https://github.com/rapidsai/cudf/pull/6051#discussion_r481807727

File: cpp/include/cudf/io/orc.hpp (hunk 20 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/orc.hpp:819]: Shouldn't be a problem

File: cpp/include/cudf/io/detail/orc.hpp [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+ */
+
+#pragma once
+
+#include <cudf/io/orc.hpp>
+
+namespace cudf {
+namespace io {
+namespace detail {
+namespace orc {
```
[the remark this answers is not in the document]
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/detail/orc.hpp:29]: I would go with another follow-up PR to fix all options, as already json and parquet have been committed.

File: cpp/include/cudf/io/orc.hpp (hunk 15 above)
Comment (cwharris) [reply cpp/include/cudf/io/orc.hpp:21]: non-standard with respect to the rest of cudf, and generally speaking keeping a filename _in_ a file is asking for inconsistency. Why should it remain?

Comment (rgsl888prabhu) [reply cpp/include/cudf/io/orc.hpp:21]: will remove it.

File: cpp/include/cudf/io/detail/orc.hpp (hunk 13 above)
Comment (rgsl888prabhu) [reply cpp/include/cudf/io/detail/orc.hpp:46]: I would go with a follow-up PR, as it is mostly related to detail API rather than options.

**Review by harrism — APPROVED** [review]

**Review by cwharris — APPROVED** [review]
lgtm :rocket: sorry for delay

**Comment by cwharris** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by kkraus14** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
rerun tests
